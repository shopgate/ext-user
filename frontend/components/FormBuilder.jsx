import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import Portal from '@shopgate/pwa-common/components/Portal';
import I18n from '@shopgate/pwa-common/components/I18n';
import TextField from '@shopgate/pwa-ui-shared/TextField';
import Button from '@shopgate/pwa-ui-shared/Button';
import RippleButton from '@shopgate/pwa-ui-shared/RippleButton';
import Select from '@shopgate/pwa-ui-shared/Form/Select';
import Checkbox from '@shopgate/pwa-ui-shared/Form/Checkbox';
import * as portals from '@shopgate/user/constants/Portals';

/**
 * Takes a form configuration and handles
 */
class FormBuilder extends Component {
  static propTypes = {
    className: PropTypes.string.isRequired,
    config: PropTypes.shape().isRequired,
    handleUpdate: PropTypes.func.isRequired,
    locale: PropTypes.string.isRequired,
    defaults: PropTypes.shape(),
  }

  static defaultProps = {
    defaults: {},
  }

  /**
   * Initializes the component.
   * @param {Object} props The components props.
   */
  constructor(props) {
    super(props);

    this.state.formData = this.props.defaults;

    this.formElements = this.buildFormElements(this.props.config, this.props.locale);
  }

  /**
   * Takes a list of which elements to render based on the respective element type
   * @param {Object} formConfig Configuration of which form fields to render
   * @param {string} locale The app's current locale setting
   * @return {Object[]}
   */
  buildFormElements(formConfig, locale) {
    const elementList = [];

    // Add all non-custom attributes and mark them as such
    Object.getOwnPropertyNames(formConfig.fields).forEach((id) => {
      if (id !== 'customAttributes') {
        const field = {
          id,
          ...formConfig.fields[id],
          custom: false,
        };
        elementList.push(field);
      }
    });

    // Add custom attributes to the list
    if (formConfig.fields.customAttributes) {
      Object.getOwnPropertyNames(formConfig.fields.customAttributes).forEach((id) => {
        const field = {
          id,
          ...formConfig.fields.customAttributes[id],
          custom: true,
        };
        elementList.push(field);
      });
    }

    // Generate handler functions for each element
    elementList.map(element => ({
      ...element,
      handleChange: (value) => {
        const newState = { ...this.state.formData };
        newState[element.id] = value;
        this.setState(newState);

        this.props.handleUpdate(this.state.formData);
      },
    }));

    // Add translation strings to the form elements, where te've been assigned by config
    if (locale) {
      // TODO: fill up with translation here using "locale" parameter
      // TODO: Make sure to fall back to defaults if possible
    }

    // Sort the fields by sortOrder property and return the full list
    return elementList.sort((element1, element2) => {
      // Keep relative sort order when no specific sort order was set for both
      if (element2.sortOrder === undefined) {
        return -1;
      } else if (element1.sortOrder === undefined) {
        return 1;
      }

      // Sort in ascending order of sortOrder otherwise
      return element1.sortOrder - element2.sortOrder;
    });
  }

  renderElement = (/* element */) => {
    return (
      <Fragment>
        {/* Nothing to render, yet */}
      </Fragment>
    );
  }

  /**
   * Renders
   * @return {JSX}
   */
  render() {
    return (
      <Fragment>
        <div className={this.props.className}>
          {this.formElements.forEach(element => this.renderElement(element))}
        </div>
      </Fragment>
    );
  }
}

export default FormBuilder;
