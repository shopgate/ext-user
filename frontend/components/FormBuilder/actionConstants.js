export const ACTION_TYPE_UPDATE_PROVINCE_ELEMENT = 'updateProvinceElement';
export const ACTION_TYPE_SET_VISIBILITY = 'setVisibility';
export const ACTION_TYPE_SET_VALUE = 'setValue';
export const ACTION_TYPE_TRANSFORM = 'transform';

export const ACTION_SET_VALUE_FIXED = 'fixed';
export const ACTION_SET_VALUE_COPY_FROM = 'copyFrom';
export const ACTION_SET_VALUE_LENGTH_OF = 'lengthOf';

export const ACTION_RULE_TYPE_NOT_IN = 'notIn';
export const ACTION_RULE_TYPE_ONE_OF = 'oneOf';
export const ACTION_RULE_TYPE_BOOLEAN = 'boolean';
export const ACTION_RULE_TYPE_REGEX = 'regex';

// Rule data is formatted as array
export const ACTION_RULE_DATA_TYPES = {
  [ACTION_RULE_TYPE_NOT_IN]: 'array',
  [ACTION_RULE_TYPE_ONE_OF]: 'array',
  [ACTION_RULE_TYPE_BOOLEAN]: 'boolean',
  [ACTION_RULE_TYPE_REGEX]: 'string',
};

export const ACTION_RULES_CONCAT_METHOD_ALL = 'all';
export const ACTION_RULES_CONCAT_METHOD_ANY = 'any';
export const ACTION_RULES_CONCAT_METHOD_NONE = 'none';
