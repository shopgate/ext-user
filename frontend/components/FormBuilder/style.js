import { css } from 'glamor';
import { themeConfig } from '@shopgate/pwa-common/helpers/config';

const { variables, colors } = themeConfig;

const fields = css({
  backgroundColor: colors.light,
  marginTop: 2,
  padding: variables.gap.bigger,
  paddingBottom: variables.gap.xsmall,
}).toString();

export default {
  fields,
};
