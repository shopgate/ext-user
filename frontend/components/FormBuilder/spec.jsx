import React from 'react';
import { mount, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import FormBuilder from '.';

configure({ adapter: new Adapter() });

describe('<FormBuilder />', () => {
  it('should render empty form', () => {
    const wrapper = mount((
      <FormBuilder
        config={{ fields: {} }}
        id="foo"
        handleUpdate={() => {}}
      />
    ));

    expect(wrapper).toMatchSnapshot();
  });

  it('should render two text fields', () => {
    const wrapper = mount((
      <FormBuilder
        config={{
          fields: {
            firstName: {
              label: 'foo',
              type: 'text',
              visible: true,
            },
            lastName: {
              label: 'bar',
              type: 'text',
              visible: true,
            },
          },
        }}
        id="foo"
        handleUpdate={() => {}}
      />
    ));

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('TextField').length).toEqual(2);
  });

  it('should not render invisible field', () => {
    const wrapper = mount((
      <FormBuilder
        config={{
          fields: {
            firstName: {
              label: 'foo',
              type: 'text',
              visible: false,
            },
          },
        }}
        id="foo"
        handleUpdate={() => {}}
      />
    ));

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('TextField').length).toEqual(0);
  });
});
