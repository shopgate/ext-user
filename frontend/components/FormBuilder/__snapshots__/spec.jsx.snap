// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormBuilder /> should not render invisible field 1`] = `
<FormBuilder
  className=""
  config={
    Object {
      "fields": Object {
        "firstName": Object {
          "label": "foo",
          "type": "text",
          "visible": false,
        },
      },
    }
  }
  defaults={Object {}}
  handleUpdate={[Function]}
  id="foo"
>
  <div
    className=""
  >
    <Portal
      name="nav-menu.foo.firstName.before"
      props={null}
    />
    <Portal
      name="nav-menu.foo.firstName"
      props={null}
    >
      <TextElement
        element={
          Object {
            "custom": false,
            "handleChange": [Function],
            "id": "firstName",
            "label": "foo",
            "type": "text",
            "visible": false,
          }
        }
        errorText=""
        name="foo.firstName"
        style={
          Object {
            "field": "",
          }
        }
        value=""
        visible={false}
      />
      <SelectElement
        element={
          Object {
            "custom": false,
            "handleChange": [Function],
            "id": "firstName",
            "label": "foo",
            "type": "text",
            "visible": false,
          }
        }
        errorText=""
        name="foo.firstName"
        style={
          Object {
            "fields": "",
          }
        }
        value=""
        visible={false}
      />
      <CountryElement
        countryList={
          Object {
            "AT": "Austria",
            "DE": "Germany",
            "FR": "France",
            "GB": "United Kingdom",
            "US": "United States",
          }
        }
        element={
          Object {
            "custom": false,
            "handleChange": [Function],
            "id": "firstName",
            "label": "foo",
            "type": "text",
            "visible": false,
          }
        }
        errorText=""
        name="foo.firstName"
        style={
          Object {
            "fields": "",
          }
        }
        value=""
        visible={false}
      />
      <ProvinceElement
        element={
          Object {
            "custom": false,
            "handleChange": [Function],
            "id": "firstName",
            "label": "foo",
            "type": "text",
            "visible": false,
          }
        }
        errorText=""
        name="foo.firstName"
        provincesList={Object {}}
        style={
          Object {
            "fields": "",
          }
        }
        value=""
        visible={false}
      />
      <RadioElement
        element={
          Object {
            "custom": false,
            "handleChange": [Function],
            "id": "firstName",
            "label": "foo",
            "type": "text",
            "visible": false,
          }
        }
        errorText=""
        name="foo.firstName"
        style={
          Object {
            "fields": "",
          }
        }
        value=""
        visible={false}
      />
      <CheckboxElement
        element={
          Object {
            "custom": false,
            "handleChange": [Function],
            "id": "firstName",
            "label": "foo",
            "type": "text",
            "visible": false,
          }
        }
        errorText=""
        name="foo.firstName"
        style={
          Object {
            "fields": "",
          }
        }
        value={false}
        visible={false}
      />
    </Portal>
    <Portal
      name="nav-menu.foo.firstName.after"
      props={null}
    />
  </div>
</FormBuilder>
`;

exports[`<FormBuilder /> should render empty form 1`] = `
<FormBuilder
  className=""
  config={
    Object {
      "fields": Object {},
    }
  }
  defaults={Object {}}
  handleUpdate={[Function]}
  id="foo"
>
  <div
    className=""
  />
</FormBuilder>
`;

exports[`<FormBuilder /> should render two text fields 1`] = `
<FormBuilder
  className=""
  config={
    Object {
      "fields": Object {
        "firstName": Object {
          "label": "foo",
          "type": "text",
          "visible": true,
        },
        "lastName": Object {
          "label": "bar",
          "type": "text",
          "visible": true,
        },
      },
    }
  }
  defaults={Object {}}
  handleUpdate={[Function]}
  id="foo"
>
  <div
    className=""
  >
    <Portal
      name="nav-menu.foo.firstName.before"
      props={null}
    />
    <Portal
      name="nav-menu.foo.firstName"
      props={null}
    >
      <TextElement
        element={
          Object {
            "custom": false,
            "handleChange": [Function],
            "id": "firstName",
            "label": "foo",
            "type": "text",
            "visible": true,
          }
        }
        errorText=""
        name="foo.firstName"
        style={
          Object {
            "field": "",
          }
        }
        value={null}
        visible={true}
      >
        <TextField
          className=""
          errorText=""
          hintText=""
          label="foo"
          multiLine={false}
          name="foo.firstName"
          onChange={[Function]}
          onFocusChange={[Function]}
          onSanitize={[Function]}
          onValidate={[Function]}
          password={false}
          setRef={[Function]}
          type="text"
          value={null}
        >
          <div
            className="css-n6iwau "
          >
            <Hint
              hintText=""
              visible={false}
            >
              <div
                className="css-tnfb18 css-1jyyyen"
              >
                <Translate
                  className=""
                  params={Object {}}
                  string=""
                >
                  <span
                    className=""
                  />
                </Translate>
              </div>
            </Hint>
            <Label
              hasErrorMessage={false}
              isFloating={false}
              isFocused={false}
              label="foo"
              name="foo.firstName"
            >
              <label
                className="css-u7lqu6 css-1fekhao"
                htmlFor="foo.firstName"
              >
                <Translate
                  className=""
                  params={Object {}}
                  string="foo"
                >
                  <span
                    className=""
                  >
                    foo
                  </span>
                </Translate>
              </label>
            </Label>
            <FormElement
              id="foo.firstName"
              multiLine={false}
              name="foo.firstName"
              onChange={[Function]}
              onFocusChange={[Function]}
              onSanitize={[Function]}
              onValidate={[Function]}
              password={false}
              setRef={[Function]}
              type="text"
              value={null}
            >
              <Component
                className="css-keangt"
                id="foo.firstName"
                multiLine={false}
                name="foo.firstName"
                onChange={[Function]}
                onFocusChange={[Function]}
                onSanitize={[Function]}
                onValidate={[Function]}
                password={false}
                setRef={[Function]}
                type="text"
                validateOnBlur={true}
                value={null}
              >
                <SimpleInput
                  autoComplete={false}
                  autoCorrect={false}
                  className="css-keangt"
                  disabled={false}
                  id="foo.firstName"
                  multiLine={false}
                  name="foo.firstName"
                  onChange={[Function]}
                  onFocusChange={[Function]}
                  onSanitize={[Function]}
                  onValidate={[Function]}
                  password={false}
                  setRef={[Function]}
                  type="text"
                  validateOnBlur={true}
                  value={null}
                >
                  <input
                    autoComplete="off"
                    autoCorrect="off"
                    className="css-keangt"
                    disabled={false}
                    id="foo.firstName"
                    name="foo.firstName"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    type="text"
                    value=""
                  />
                </SimpleInput>
              </Component>
            </FormElement>
            <Underline
              hasErrorMessage={false}
              isFocused={false}
            >
              <div
                className="css-ycal4l"
              >
                <div
                  className="css-1tpmuef"
                  style={
                    Object {
                      "borderBottomColor": undefined,
                      "transform": "scale3d(0,1,1)",
                    }
                  }
                />
              </div>
            </Underline>
            <ErrorText
              className=""
              errorText=""
              validationError={null}
            >
              <div
                className=" css-hmgfge"
              >
                <Translate
                  className=""
                  params={Object {}}
                  string=""
                >
                  <span
                    className=""
                  />
                </Translate>
              </div>
            </ErrorText>
          </div>
        </TextField>
      </TextElement>
      <SelectElement
        element={
          Object {
            "custom": false,
            "handleChange": [Function],
            "id": "firstName",
            "label": "foo",
            "type": "text",
            "visible": true,
          }
        }
        errorText=""
        name="foo.firstName"
        style={
          Object {
            "fields": "",
          }
        }
        value={null}
        visible={true}
      />
      <CountryElement
        countryList={
          Object {
            "AT": "Austria",
            "DE": "Germany",
            "FR": "France",
            "GB": "United Kingdom",
            "US": "United States",
          }
        }
        element={
          Object {
            "custom": false,
            "handleChange": [Function],
            "id": "firstName",
            "label": "foo",
            "type": "text",
            "visible": true,
          }
        }
        errorText=""
        name="foo.firstName"
        style={
          Object {
            "fields": "",
          }
        }
        value={null}
        visible={true}
      />
      <ProvinceElement
        element={
          Object {
            "custom": false,
            "handleChange": [Function],
            "id": "firstName",
            "label": "foo",
            "type": "text",
            "visible": true,
          }
        }
        errorText=""
        name="foo.firstName"
        provincesList={Object {}}
        style={
          Object {
            "fields": "",
          }
        }
        value={null}
        visible={true}
      />
      <RadioElement
        element={
          Object {
            "custom": false,
            "handleChange": [Function],
            "id": "firstName",
            "label": "foo",
            "type": "text",
            "visible": true,
          }
        }
        errorText=""
        name="foo.firstName"
        style={
          Object {
            "fields": "",
          }
        }
        value={null}
        visible={true}
      />
      <CheckboxElement
        element={
          Object {
            "custom": false,
            "handleChange": [Function],
            "id": "firstName",
            "label": "foo",
            "type": "text",
            "visible": true,
          }
        }
        errorText=""
        name="foo.firstName"
        style={
          Object {
            "fields": "",
          }
        }
        value={null}
        visible={true}
      />
    </Portal>
    <Portal
      name="nav-menu.foo.firstName.after"
      props={null}
    />
    <Portal
      name="nav-menu.foo.lastName.before"
      props={null}
    />
    <Portal
      name="nav-menu.foo.lastName"
      props={null}
    >
      <TextElement
        element={
          Object {
            "custom": false,
            "handleChange": [Function],
            "id": "lastName",
            "label": "bar",
            "type": "text",
            "visible": true,
          }
        }
        errorText=""
        name="foo.lastName"
        style={
          Object {
            "field": "",
          }
        }
        value=""
        visible={true}
      >
        <TextField
          className=""
          errorText=""
          hintText=""
          label="bar"
          multiLine={false}
          name="foo.lastName"
          onChange={[Function]}
          onFocusChange={[Function]}
          onSanitize={[Function]}
          onValidate={[Function]}
          password={false}
          setRef={[Function]}
          type="text"
          value=""
        >
          <div
            className="css-n6iwau "
          >
            <Hint
              hintText=""
              visible={false}
            >
              <div
                className="css-tnfb18 css-1jyyyen"
              >
                <Translate
                  className=""
                  params={Object {}}
                  string=""
                >
                  <span
                    className=""
                  />
                </Translate>
              </div>
            </Hint>
            <Label
              hasErrorMessage={false}
              isFloating={false}
              isFocused={false}
              label="bar"
              name="foo.lastName"
            >
              <label
                className="css-u7lqu6 css-1fekhao"
                htmlFor="foo.lastName"
              >
                <Translate
                  className=""
                  params={Object {}}
                  string="bar"
                >
                  <span
                    className=""
                  >
                    bar
                  </span>
                </Translate>
              </label>
            </Label>
            <FormElement
              id="foo.lastName"
              multiLine={false}
              name="foo.lastName"
              onChange={[Function]}
              onFocusChange={[Function]}
              onSanitize={[Function]}
              onValidate={[Function]}
              password={false}
              setRef={[Function]}
              type="text"
              value=""
            >
              <Component
                className="css-keangt"
                id="foo.lastName"
                multiLine={false}
                name="foo.lastName"
                onChange={[Function]}
                onFocusChange={[Function]}
                onSanitize={[Function]}
                onValidate={[Function]}
                password={false}
                setRef={[Function]}
                type="text"
                validateOnBlur={true}
                value=""
              >
                <SimpleInput
                  autoComplete={false}
                  autoCorrect={false}
                  className="css-keangt"
                  disabled={false}
                  id="foo.lastName"
                  multiLine={false}
                  name="foo.lastName"
                  onChange={[Function]}
                  onFocusChange={[Function]}
                  onSanitize={[Function]}
                  onValidate={[Function]}
                  password={false}
                  setRef={[Function]}
                  type="text"
                  validateOnBlur={true}
                  value=""
                >
                  <input
                    autoComplete="off"
                    autoCorrect="off"
                    className="css-keangt"
                    disabled={false}
                    id="foo.lastName"
                    name="foo.lastName"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    type="text"
                    value=""
                  />
                </SimpleInput>
              </Component>
            </FormElement>
            <Underline
              hasErrorMessage={false}
              isFocused={false}
            >
              <div
                className="css-ycal4l"
              >
                <div
                  className="css-1tpmuef"
                  style={
                    Object {
                      "borderBottomColor": undefined,
                      "transform": "scale3d(0,1,1)",
                    }
                  }
                />
              </div>
            </Underline>
            <ErrorText
              className=""
              errorText=""
              validationError={null}
            >
              <div
                className=" css-hmgfge"
              >
                <Translate
                  className=""
                  params={Object {}}
                  string=""
                >
                  <span
                    className=""
                  />
                </Translate>
              </div>
            </ErrorText>
          </div>
        </TextField>
      </TextElement>
      <SelectElement
        element={
          Object {
            "custom": false,
            "handleChange": [Function],
            "id": "lastName",
            "label": "bar",
            "type": "text",
            "visible": true,
          }
        }
        errorText=""
        name="foo.lastName"
        style={
          Object {
            "fields": "",
          }
        }
        value=""
        visible={true}
      />
      <CountryElement
        countryList={
          Object {
            "AT": "Austria",
            "DE": "Germany",
            "FR": "France",
            "GB": "United Kingdom",
            "US": "United States",
          }
        }
        element={
          Object {
            "custom": false,
            "handleChange": [Function],
            "id": "lastName",
            "label": "bar",
            "type": "text",
            "visible": true,
          }
        }
        errorText=""
        name="foo.lastName"
        style={
          Object {
            "fields": "",
          }
        }
        value=""
        visible={true}
      />
      <ProvinceElement
        element={
          Object {
            "custom": false,
            "handleChange": [Function],
            "id": "lastName",
            "label": "bar",
            "type": "text",
            "visible": true,
          }
        }
        errorText=""
        name="foo.lastName"
        provincesList={Object {}}
        style={
          Object {
            "fields": "",
          }
        }
        value=""
        visible={true}
      />
      <RadioElement
        element={
          Object {
            "custom": false,
            "handleChange": [Function],
            "id": "lastName",
            "label": "bar",
            "type": "text",
            "visible": true,
          }
        }
        errorText=""
        name="foo.lastName"
        style={
          Object {
            "fields": "",
          }
        }
        value=""
        visible={true}
      />
      <CheckboxElement
        element={
          Object {
            "custom": false,
            "handleChange": [Function],
            "id": "lastName",
            "label": "bar",
            "type": "text",
            "visible": true,
          }
        }
        errorText=""
        name="foo.lastName"
        style={
          Object {
            "fields": "",
          }
        }
        value=""
        visible={true}
      />
    </Portal>
    <Portal
      name="nav-menu.foo.lastName.after"
      props={null}
    />
  </div>
</FormBuilder>
`;
